mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs

```diff
@@ -0,0 +1,872 @@
+        };
+
+        let (connection_ready_signal, wait_for_connection_ready) = async_oneshot::channel();
+
+        let connection_loop = {
+            let last_response = Arc::new(AtomicU64::new(now_ms()));
+            let no_connection_timeout_f = {
+                let last_response = last_response.clone();
+                async move {
+                    loop {
+                        Timer::sleep(CUTOFF_TIMEOUT).await;
+                        let last_sec = (last_response.load(AtomicOrdering::Relaxed) / 1000) as f64;
+                        if now_float() - last_sec > CUTOFF_TIMEOUT {
+                            break ElectrumConnectionErr::Temporary(format!(
+                                "Server didn't respond for too long ({}s).",
+                                now_float() - last_sec
+                            ));
+                        }
+                    }
+                }
+            };
+            let mut no_connection_timeout_f = Box::pin(no_connection_timeout_f).fuse();
+
+            let recv_f = {
+                let address = address.clone();
+                let connection = connection.clone();
+                let event_handlers = event_handlers.clone();
+                async move {
+                    // FIXME: `process_electrum_response` would be better off receiving bytes and not raw json,
+                    // this is to avoid converting the json back to bytes to pass to the event handlers. Check if
+                    // this is possible.
+                    while let Some(response) = transport_rx.next().await {
+                        match response {
+                            Ok(raw_json) => {
+                                last_response.store(now_ms(), AtomicOrdering::Relaxed);
+                                connection.process_electrum_response(raw_json, &event_handlers).await;
+                            },
+                            Err(e) => {
+                                error!("{address} error: {e:?}");
+                            },
+                        }
+                    }
+                    ElectrumConnectionErr::Temporary("Receiver disconnected".to_string())
+                }
+            };
+            let mut recv_f = Box::pin(recv_f).fuse();
+
+            let (tx, rx) = mpsc::channel(0);
+            let send_f = {
+                let address = address.clone();
+                let event_handlers = event_handlers.clone();
+                let mut rx = rx_to_stream(rx).compat();
+                async move {
+                    while let Some(Ok(bytes)) = rx.next().await {
+                        // FIXME: Also here we don't need to convert the bytes to json.
+                        // That's too early considering it will be send over the network.
+                        // Fix `transport_tx.send` so that it accepts bytes instead.
+                        let raw_json: Json = match json::from_slice(&bytes) {
```

Comment:
I guess with a special type we can keep bytes and json version of the value so we never need to convert them back-and-forth again and again.

Reply:
re-reading the comment here, the word "too early" seems irrelevant.
I mean here that the `tx` side of the channel generated a jsonrpc request and serialized it into `Vec<u8>` and sent it to `rx`.

There is no reason for us to de-serialize that request. We should just send it over the network directly.

Reply:
Done in https://github.com/KomodoPlatform/komodo-defi-framework/pull/1966/commits/c9b669f611fe69522babb2b914d8867be3b5575f